src/event.cpp

```diff
@@ -140,13 +154,16 @@ void event_serializer::serialize(const memory::vector<event> &events, ddwaf_resu
             output.data = to_cstr(buffer);
         }
 
-        if (!actions.empty()) {
+        if (!all_actions.empty()) {
             // NOLINTNEXTLINE
-            output.actions.array = static_cast<char **>(malloc(sizeof(char *) * actions.size()));
-            output.actions.size = actions.size();
+            output.actions.array =
+                static_cast<char **>(malloc(sizeof(char *) * all_actions.size()));
+            output.actions.size = all_actions.size();
 
             std::size_t index = 0;
-            for (const auto &action : actions) { output.actions.array[index++] = to_cstr(action); }
+            for (const auto &action : all_actions) {
+                output.actions.array[index++] = to_cstr(action);
```

Comment:
you could save some allocations here by not using to_cstr() by adding to all_actions a string_view created from std::string.c_str(), using the std::string in rule.actions

Reply:
As you mentioned privately, these need to outlive the ruleset.